modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/statechannel/FeeCalculator.scala

```diff
@@ -0,0 +1,97 @@
+)
+
+object FeeCalculatorConfig {
+  val noFee: FeeCalculatorConfig =
+    FeeCalculatorConfig(baseFee = 0L, stakingWeight = BigDecimal(0.0), computationalCost = 1L, proWeight = BigDecimal(0.0))
+
+  def getByOrdinal(configs: SortedMap[SnapshotOrdinal, FeeCalculatorConfig])(ordinal: SnapshotOrdinal): FeeCalculatorConfig =
+    configs
+      .rangeTo(ordinal)
+      .lastOption
+      .map { case (_, config) => config }
+      .getOrElse(FeeCalculatorConfig.noFee)
+}
+
+sealed trait FeeCalculator[F[_]] {
+  def calculateRecommendedFee(
+    maybeOrdinal: Option[SnapshotOrdinal],
+    delay: NonNegLong = 0L
+  )(staked: Balance, sizeKb: NonNegInt, tip: NonNegLong = 0L, proScore: FeeCalculator.ProScore = 0.0d): F[SnapshotFee]
+}
+
+object FeeCalculator {
+  def make[F[_]: MonadThrow](configs: SortedMap[SnapshotOrdinal, FeeCalculatorConfig]): FeeCalculator[F] =
+    new FeeCalculator[F] {
+      private val allConfigs: NonEmptyList[FeeCalculatorConfig] =
+        NonEmptyList
+          .fromList(configs.values.toList)
+          .getOrElse(NonEmptyList.one(FeeCalculatorConfig.noFee))
+
+      private def getConfig = FeeCalculatorConfig.getByOrdinal(configs)(_)
+
+      /** Calculates the recommended fee for given stake, size and pro score and adds optional tip on top. Calculates the worst case
+        * scenario (highest) fee based on the provided configs range picked as follows:
+        *   - all configs when ordinal is None
+        *   - ordinal to ordinal + delay when ordinal is non empty.
+        */
+      def calculateRecommendedFee(
+        maybeOrdinal: Option[SnapshotOrdinal],
+        delay: NonNegLong
+      )(staked: Balance, sizeKb: NonNegInt, tip: NonNegLong, proScore: ProScore): F[SnapshotFee] =
+        maybeOrdinal
+          .fold(allConfigs) { ordinal =>
+            NonEmptySet
+              .of(ordinal, ordinal.plus(delay))
+              .toNonEmptyList
+              .map(getConfig)
+          }
+          .traverse(calculate(_)(staked, sizeKb, tip, proScore))
+          .map(_.maximum)
+    }
+
+  private def calculate[F[_]: MonadThrow](
+    config: FeeCalculatorConfig
+  )(staked: Balance, sizeKb: NonNegInt, tip: NonNegLong, proScore: ProScore): F[SnapshotFee] = {
+    val FeeCalculatorConfig(baseFee, stakingWeight, computationalCost, proWeight) = config
+    val workAmount: BigDecimal = BigDecimal(sizeKb) * BigDecimal(computationalCost)
+
+    val workMultiplier: BigDecimal =
+      BigDecimal(1) / (BigDecimal(1) + BigDecimal(staked.value) * stakingWeight + BigDecimal(proScore) * proWeight)
+
+    val roundUp = new MathContext(0, RoundingMode.UP)
+    val fee = (BigDecimal(baseFee) * workAmount * workMultiplier + BigDecimal(tip)).round(roundUp).toBigInt
+
+    Either
+      .cond(fee.isValidLong, fee.longValue, "calculated fee exceeded long max value")
+      .flatMap(NonNegLong.from)
+      .map(SnapshotFee(_))
+      .leftMap(msg => new Error(s"Critical error during snapshot fee calculation: $msg!"))
```

Comment:
```suggestion
      .leftMap(msg => new IllegalStateException(s"Critical error during snapshot fee calculation: $msg!"))
```

I don't think we should be creating `java.lang.Error`

Reply:
I can change it, but can you elaborate why we shouldn't?

Reply:
I used `Error` as it is an unhandleable situation - we can't recover from it.

Reply:
> I can change it, but can you elaborate why we shouldn't?

Same reason why I don't feel we should new `Throwable`.
My feeling has been `java.lang.Error` is reserved for the JVM whereas `Exception` is for us mere mortals.
I'm open to a more modern explanation/approach though...

Reply:
Ok I did some more reading and I agree it seems `Error` is for more severe and mainly architectural stuff, like `StackOverflow` and such. Hence using `IllegalStateException` seems more appropriate given it's still an unchecked exception.